odl/discr/partition.py

```diff
@@ -0,0 +1,561 @@
+
+    def insert(self, index, other):
+        """Return a copy with ``other`` inserted before ``index``.
+
+        Parameters
+        ----------
+        index : `int`
+            Index of the dimension before which ``other`` is to
+            be inserted. Must fulfill ``0 <= index <= ndim``.
+        other : `RectPartition`
+            Partition to be inserted
+
+        Returns
+        -------
+        newpart : `RectPartition`
+            Partition with the inserted other partition
+
+        Examples
+        --------
+        """
+        newgrid = self.grid.insert(index, other.grid)
+        newbbox = self.bbox.insert(index, other.bbox)
+        new_b = newbbox.begin
+        new_e = newbbox.end
+        return RectPartition(newgrid, begin=new_b, end=new_e)
+
+    # TODO: pretty-print
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        inner_str = '\n {!r}'.format(self.grid)
+        sep = ',\n '
+        if self._custom_begin:
+            inner_str += sep + 'begin={}'.format(
+                array1d_repr(self.bbox.begin))
+            sep = ', '
+        if self._custom_end:
+            inner_str += sep + 'end={}'.format(
+                array1d_repr(self.bbox.end))
+
+        return '{}({})'.format(self.__class__.__name__, inner_str)
+
+
+def uniform_partition(intv_prod, num_nodes, nodes_on_bdry=True):
+    """Return a partition of ``intv_prod`` by a regular grid.
+
+    Parameters
+    ----------
+    intv_prod : `IntervalProd`
+        Interval product to be partitioned
+    num_nodes : `int` or sequence of `int`
+        Number of nodes per axis. For 1d intervals, a single integer
+        can be specified.
+    nodes_on_bdry : `bool` or boolean array-like
+        If `True`, place the outermost grid points at the boundary. For
+        `False`, they are shifted by half a cell size to the 'inner'.
+        If an array-like is given, it must have shape ``(ndim, 2)``,
+        where ``ndim`` is the number of dimensions. It defines per axis
+        whether the leftmost (first column) and rightmost (second column)
+        nodes node lie on the boundary.
+
+    Examples
+    --------
+    """
+    # Sanity checks
+    if np.shape(num_nodes) == ():
+        num_nodes = (num_nodes,)
+    elif len(num_nodes) != intv_prod.ndim:
+        raise ValueError('num_nodes has length {}, expected {}.'
+                         ''.format(len(num_nodes), (intv_prod.ndim)))
+
+    if np.shape(nodes_on_bdry) == ():
+        nodes_on_bdry = ([(bool(nodes_on_bdry), bool(nodes_on_bdry))] *
+                         intv_prod.ndim)
+    elif np.shape(nodes_on_bdry) == (intv_prod.ndim, 2):
+        pass
+    else:
+        raise ValueError('node_at_bdry has shape {}, expected {}.'
+                         ''.format(np.shape(nodes_on_bdry),
+                                   (intv_prod.ndim, 2)))
+
+    gmin, gmax = [], []
+    for n, beg, end, (bdry_l, bdry_r) in zip(num_nodes, intv_prod.begin,
```

Comment:
It's actually not that complicated, maybe just badly explained. It's just a matter of where to place the min and max grid points so the final grid is a regular grid.